sync/sync.go

```diff
@@ -3,79 +3,127 @@ package sync
-		return errors.New("synchronizer is stopped")
+func (s *Synchronizer) fetcherTask(ctx context.Context, height uint64, verifiers *stream.Stream, errChan chan ErrSyncFailed) stream.Callback {
+	for {
+		select {
+		case <-ctx.Done():
+			return func() {}
+		default:
+			block, err := s.StarknetData.BlockByNumber(ctx, height)
+			if err != nil {
+				continue
+			}
+			stateUpdate, err := s.StarknetData.StateUpdate(ctx, height)
+			if err != nil {
+				continue
+			}
+
+			return func() {
+				verifiers.Go(func() stream.Callback { return s.verifierTask(ctx, block, stateUpdate, errChan) })
+			}
+		}
 	}
-	close(s.quit)
-	return nil
 }
 
-func (s *Synchronizer) SyncBlocks() error {
-	for {
+func (s *Synchronizer) verifierTask(ctx context.Context, block *core.Block, stateUpdate *core.StateUpdate, errChan chan ErrSyncFailed) stream.Callback {
+	err := s.Blockchain.SanityCheckNewHeight(block, stateUpdate)
+	return func() {
 		select {
-		case <-s.quit:
-			// Clean up and return
-			return nil
+		case <-ctx.Done():
+			return
 		default:
-			// Call the gateway to get blocks and state update
-			nextHeight := uint64(0)
-			if h, err := s.Blockchain.Height(); err == nil {
-				nextHeight = h + 1
-			}
-			block, err := s.StarknetData.BlockByNumber(context.Background(), nextHeight)
 			if err != nil {
-				return err
+				s.log.Warnw("Sanity checks failed", "number", block.Number, "hash", block.Hash.Text(16))
+				errChan <- ErrSyncFailed{block.Number, err}
+				return
 			}
-			s.log.Infow("Fetched block", "number", block.Number, "hash", block.Hash.Text(16))
-			stateUpdate, err := s.StarknetData.StateUpdate(context.Background(), nextHeight)
+			err = s.Blockchain.Store(block, stateUpdate)
 			if err != nil {
-				return err
+				s.log.Warnw("Failed storing Block", "number", block.Number, "hash", block.Hash.Text(16),
+					"err", err.Error())
+				errChan <- ErrSyncFailed{block.Number, err}
+				return
 			}
-			s.log.Infow("Fetched state update", "newRoot", stateUpdate.NewRoot.Text(16), "hash", stateUpdate.BlockHash.Text(16))
-			if err = s.Blockchain.Store(block, stateUpdate); err != nil {
-				return err
+
+			s.log.Infow("Stored Block", "number", block.Number, "hash", block.Hash.Text(16),
+				"root", block.GlobalStateRoot.Text(16))
+		}
+	}
+}
+
+func (s *Synchronizer) SyncBlocks(syncCtx context.Context) error {
+	errChan := make(chan ErrSyncFailed)
+	fetchers := stream.New().WithMaxGoroutines(runtime.NumCPU())
+	verifiers := stream.New().WithMaxGoroutines(runtime.NumCPU())
+
+	streamCtx, streamCancel := context.WithCancel(syncCtx)
+	nextHeight := uint64(0)
+	if h, err := s.Blockchain.Height(); err == nil {
+		nextHeight = h + 1
+	}
+
+	for {
+		select {
+		case err := <-errChan:
+			streamCancel() // cancel all running tasks
+			streamCtx, streamCancel = context.WithCancel(syncCtx)
+			nextHeight = err.Height // keep syncing from failed height
+			s.log.Warnw("Rolling back sync process to failed height", "height", err.Height)
```

Comment:
Quick question regarding this. Do we at any point stop restarting due to too many errors? Or we just keep restarting until the user stops the sync?

Reply:
We never stop, that is supposed to be the expected behaviour #568. However, if there are any panics in the goroutines they only are propagated up when the user sends the signal to stop. Now, this may or may not be a desired behaviour. Also, in general, panics are not detected by the parent go routine.

Reply:
the latter